sklearn/tree/_splitter.pyx

```diff
@@ -334,16 +349,36 @@ cdef class BestSplitter(BaseDenseSplitter):
             # f_j in the interval [n_total_constants, f_i[
             current.feature = features[f_j]
 
-            # Sort samples along that feature; by
-            # copying the values into an array and
+            n_missing = 0
+            # Sort samples along that feature; by copying the values into an array and
             # sorting the array in a manner which utilizes the cache more
             # effectively.
-            for i in range(start, end):
-                Xf[i] = self.X[samples[i], current.feature]
+            # Missing values are placed at the end and do not participate in the sorting.
+            if has_missings[current.feature]:
+                i, j = start, end - 1
+                while i <= j:
```

Comment:
In this loop we swap values from the "start" of the array with values at the "end". This makes me think that the values won't be in the cache because it is hard to know for the CPU what we will look at next. And the code needs a bit of paying attention to understand it.

Overall it makes me wonder if it would be equally as fast (or faster) to first find where the lowest value of `j` by iterating once from the end of `self.X` to the first not missing value. And then iterate from the front of `X` and move missing values to the "end".

Maybe you've already tried something like this or know something I don't know that explains why it makes sense to have a loop like this. I'd love to know (and add it as a comment for readers from the future who might have the same question).

Reply:
This part does not actually take advantage of the cache. It is setting up `Xf`, so that `sort` can be called on a contiguous array.

> Overall it makes me wonder if it would be equally as fast (or faster) to first find where the lowest value

I think this is what the current algorithm does. Although, I would be happy if there was a way to simplify the logic. Here is the code with comments:

```python
while i <= j:
    # Finds the lowest j that is not missing
    if isnan(self.X[samples[j], current.feature]):
        n_missing += 1
        j -= 1
        continue

    # X[samples[j]] is a non-missing value
    if isnan(self.X[samples[i], current.feature]):
        # If X[samples[i]] is missing, swap samples[i] and samples[j]
        samples[i], samples[j] = samples[j], samples[i]
        n_missing += 1
        j -= 1

    Xf[i] = self.X[samples[i], current.feature]
    i += 1
```

When a missing value is moved to the "end", the "end" needs to be updated with the next valid end position. This is what `j` is keeping track of.

Technically, the algorithm can be simplified to this:

```python
while i <= j:
    if isnan(self.X[samples[j], current.feature]):
        n_missing += 1
        j -= 1
        continue

    if isnan(self.X[samples[i], current.feature]):
        samples[i], samples[j] = samples[j], samples[i]

    Xf[i] = self.X[samples[i], current.feature]
    i += 1
```

The additional `j-=1` in the second `if` is an optimization to not run `isnan` again when finding the next `j`.

I updated the source with more comments to explain the logic.

Reply:
Thanks for the explanations! Originally what I had in mind was something like the following, as a way of disentangling the `while` loop:
```python
# simplified with
Xf = [1, 42, 23, 18, np.nan, 12, np.nan, np.nan]
# Find the "actual" end of the feature vector
for j in reversed(range(len(Xf)-1)):
    if np.isnan(Xf[j]):
        current_end = j

# move NaNs to the end
for i in range(current_end):
    if i >= current_end:
        break
    if np.isnan(Xf[i]):
        Xf[i], Xf[current_end-1] = Xf[current_end-1], Xf[i]
        current_end -= 1
```

Reply:
Hmm, I do not think that works with input such as:

```python
Xf = [1, 12, np.nan, 1]
```

The PR's implementation is keeping track of "two pointers" and moving them around until they meet. To me, that calls for a `while` loop. 

(This feels like one of those software engineering interview questions 😅)

Reply:
I do like the use of `current_end` instead of `j`. I updated the PR with this change here: [`3a05ce5` (#23595)](https://github.com/scikit-learn/scikit-learn/pull/23595/commits/3a05ce522d930e74cd521fcadfb485c0c6f9cd07)

Reply:
I hadn't thought of `[1, 12, np.nan, 1]`. I now vote for keeping the `while` loop.

Good thing I have a job and don't need to pass a coding interview right now :D